### ISSUE TITLE
[no-use-before-define] False positive when using the `typeof` keyword

### ISSUE THREAD

**Issue by manuth** [typescript-eslint/typescript-eslint#2572]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-use-before-define": [
        "warn",
        {
            "classes": false,
            "ignoreTypeReferences": true
        }
    ]
  }
}
```

```TS
interface QuestionOptions
{
    type: typeof MyQuestion.TypeName;
}

class MyQuestion
{
    public static readonly TypeName = "my-custom-question";
}
```

**Expected Result**
No error should be reported as classes and type-references are ignored.

**Actual Result**
An error is reported:
```
'MyQuestion' was used before it was defined. (@typescript-eslint/no-use-before-define)
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.1.2-alpha.3` |
| `@typescript-eslint/parser`        | `4.1.1` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.9.0` |
| `node`                             | `12.18.3` |

**Comment by Drakota** [comment]
Hi!
I'd like to take on this issue if it's not taken.

I've checked quickly and I think the issue is with the `reference.isTypeReference` being false (The reference takes MyQuestion instead of TypeName).

The classes option on the other hand seems to be working as intented per the [documentation](https://eslint.org/docs/rules/no-use-before-define#classes). Both declarations are in the Global scope like this:
```ts
// this is incorrect
new A();
class A {
}
```

Any insights on this?

**Comment by bradzacher** [comment]
Your reasoning is in the right vein! But not quite right because `typeof` is a bit tricky.

The thing that's passed to the TS `typeof` operator is actually a ***value*** reference.
This is intentional because you cannot reference _types_ in the argument position of TS's `typeof` operator, only _values_.

For example:
```ts
const c = 2;
type TestC = typeof c; // === 2

type T = 1;
type TestT = typeof T; // 'T' only refers to a type, but is being used as a value here.
```
[repl](https://www.typescriptlang.org/play?#code/MYewdgzgLgBMMF4YCYDcAoKBPADgUxgBU9oBhRGbfEAMzlRgHpHEEll1NcDCKBGDFR4kovJENpEGzGAHJCsmOAA2WGACc8NPOoiUQMAIaVuAGhgAjAK6wAlnot5bYAOYwrEPABMje4wDdDZSsCAAsdPAA6dCA)

So if we were to set `reference.isTypeReference`, then `typeof T` would incorrectly resolve to the type `T`, and it would never be able to resolve to the variable `c`.
So the scope analyser is working as expected here.

Instead, what's not working as expected is the rule. The rule currently makes the assumption that the only "type references" are references with `isTypeReference === true`. Whilst this is correct in the strictest sense, it's not correct from the POV of a user; users think of TS's `typeof` operator as being a type reference.

Does that help?

**Comment by manuth** [comment]
That said, I didn't mean to confuse anyone by disabling `ignoreTypeReferences`
I just tried to set every option which might help to make my code working.

**Comment by bradzacher** [comment]
there's nothing wrong with your options or thinking!
This was just a case that I didn't think about when I updated the rule for the v4 release.

**Comment by Drakota** [comment]
As users we think of TS's `typeof` to return a type reference, which we think the option `ignoreTypeReferences` also refers to, but we can't set `isTypeReference` like in my example, because that would erroneously also work for a `typeof` for a Type as `typeof` only takes value references. 

So if I'm following correctly correctly we should also take into account `reference.isValueReference`, but not any value reference only the one that's been prefixed with `typeof`.

Is there a way to know if a `TSESLint.Scope.Reference` has `typeof`?

**Comment by bradzacher** [comment]
Yup you're all correct!

> Is there a way to know if a TSESLint.Scope.Reference has typeof?

`reference.identifier` will give you the identifier for the reference, and you can inspect its parent to see if it's a `TSTypeQuery`.
